design/general_refactorings.md

```diff
@@ -0,0 +1,381 @@
+        try:
+            fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+            logging.debug("Successfully fetched certificate")
+        except Exception as error:
+            raise CertificateRetrievalFailedException(f"Unable to retrieve certificate from {hostname}:{port}. " \
+                            f"Reason: {str(error)}") from error
+
+        return fingerprint, readable_certificate
+```
+
+Appendinx 1 lists the error handling cases where fixes need to be done.
+
+------
+
+### Strong-type function parameters and return types
+
+##### Proposal
+
+Strong-type all parameters and function return types as per Python3 best practices
+
+##### Current state
+
+Functions are not strongly typed.
+
+For example, `def get_variable(self, variable_data):`
+
+##### Recommendation
+
+Strongly Type the function so it will resemble `def get_variable(self, variable_data : VariableData) -> str:`"
+
+------
+
+### Explicit Import Statements
+
+##### Proposal
+
+Make sure all import statement is explicit and contains full path. Explicit paths makes it easier to tell exactly what the imported resource is and where it is. Additionally, absolute imports remain valid even if the location of the imported resources changes.
+
+##### Current state
+
+The codebase contains implicit import statements
+
+```python
+from .client import Client
+```
+
+##### Recommendation
+
+Use explicit import statements instead.
+
+```python
+from conjur.client import Client
+```
+
+
+
+## APPENDIX
+
+### APPENDIX 1: Error Handling
+
+------
+
+The following is a mapping of builtin and third-party exceptions that are currently raised in Conjur. In this document, I will provide suggestions for Conjur CLI/SDK-specific exceptions we should use instead.
+
+* api.py
+
+  * `def __init__()` - There are a few places that raise similar exceptions. `RuntimeError` can be replaced with `MissingParameterException`
+
+    * Missing Account Name
+
+      ```python
+      self._account = account
+      if not self._account:
+          raise RuntimeError("Account cannot be empty!")
+      ```
+
+    * Missing Url
+
+      * ```python
+        self._account = account
+        if not self._account:
+            raise RuntimeError("Account cannot be empty!")
+        ```
+
+  * `def login()` - `RuntimeError` can be replaced with `MissingParameterException`
+
+    * login id or password are missing 
+
+      ```python
+      if not login_id or not password:
+          # TODO: Use custom error
+          raise RuntimeError("Missing parameters in login invocation!")
+      ```
+
+  * `def authenticate()`
+
+    * login id or password are missing 
+
+      * ```python
+        if not self.login_id or not self.api_key:
+            # TODO: Use custom error
+            raise RuntimeError("Missing parameters in authentication invocation!")
+        ```
+
+  * `def rotate_other_api_key()` 
+
+    * `Exception` should be replaced with `InvalidResourceException`
```

Comment:
this is not related to the others so why is it under `RuntimeError` can be replaced with `MissingParameterException`?

Reply:
This is about this function
`    def rotate_other_api_key(self, resource: Resource):
        """
        This method is used to rotate a user/host's API key that is not the current user.
        To rotate API key of the current user use rotate_personal_api_key
        """
        if resource.type not in ('user', 'host'):
            raise Exception("Error: Invalid resource type")

        # Attach the resource type (user or host)
        query_params = {
            'role': resource.full_id()
        }
        response = invoke_endpoint(HttpVerb.PUT, ConjurEndpoint.ROTATE_API_KEY,
                                   self._default_params,
                                   api_token=self.api_token,
                                   ssl_verify=self._ssl_verify,
                                   query=query_params).text
        return response`


It raises Exception. What RuntimeError you mean?